segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/WriterTableProcessor.java

```diff
@@ -238,16 +291,16 @@ public String toString() {
         // for each such bucket and finally (reindex) update the bucket.
         return this.indexWriter
                 .groupByBucket(segment, keyUpdates.getUpdates(), timer)
-                .thenComposeAsync(bucketUpdates -> fetchExistingKeys(bucketUpdates, segment, timer)
+                .thenComposeAsync(builders -> fetchExistingKeys(builders, segment, timer)
                                 .thenComposeAsync(v -> {
-                                            logBucketUpdates(bucketUpdates);
-                                            return this.indexWriter.updateBuckets(segment, bucketUpdates,
-                                                    this.aggregator.getLastIndexedOffset(), keyUpdates.getLastIndexedOffset(),
-                                                    keyUpdates.getTotalUpdateCount(), timer.getRemaining());
-                                        },
-                                        this.executor)
-                                .thenApply(ignored -> keyUpdates.getLastIndexedOffset()),
-                        this.executor);
+                                    val bucketUpdates = builders.stream().map(BucketUpdate.Builder::build).collect(Collectors.toList());
+                                    logBucketUpdates(bucketUpdates);
+                                    return this.indexWriter.updateBuckets(segment, bucketUpdates,
+                                            this.aggregator.getLastIndexedOffset(), keyUpdates.getLastIndexedOffset(),
+                                            keyUpdates.getTotalUpdateCount(), timer.getRemaining());
```

Comment:
keyUpdateCollection is not thread safe, but this looks like it's occurring in another thread. Perhaps keyUpdates should be passes explicitly as an argument here?

Reply:
I think the best solution here is to make KeyUpdateCollection thread safe. I fixed it.